### PR TITLE
Optimize python QgsFeature.setAttribute(s) calls

### PR DESCRIPTION
**Pull request by ptitjano**
### Context
I have encountered some performance issues while working on a [python plugin](https://plugins.qgis.org/plugins/qduckdb/) to provide a [duckdb](https://duckdb.org/) provider. DuckDB is a file based database (like SQLite) which has an extension to handle spatial data. Is is well suited to handle tabular data and formats such as [parquet](https://parquet.apache.org/).

I wanted to benchmark some performance of the `duckdb` provider.     
I used some dataset of the buildings of New York City. It contains 1_607_055 geometries and 5 attributes (1 integer and 4 string). The database was saved in the duckdb and shapefile format. 
The `ogr` format needs 6 seconds to read all the features while the `duckdb` provider needs 18.2 seconds.    
I have found this result quite puzzling. Of course, I did not expect the `duckdb` provider to beat the `ogr` provider. A provider written in python could not beat come some c++ code but I would not have expected such a difference.    

The python code of the provider itself is quite simple. It just creates a SQL query  based on the QGIS request, calls duckdb to execute it and then sets features by calling QgsFeature.setId / setGeometry / setAttributes and so on. There is not much that could be achieved on the python side. 

### Some initial benchmark

So, I have tried to understand what takes most of the time. It turns out than a python provider which does absolutely nothing needs 3.4 seconds to process 1_607_055 features. I guess, this is price we have to pay because of the python overlay.    
However, I was surprised that most of the time is spent on the `QgsFeature.setAttributes` call. Some details

| operation | duration | total percentage | accumulated time |
|-----------|-------|-------------|----|
| empty loop | 3.40s | 18.72% |3.40s |
| duckdb "SELECT" request | 0.23s | 1.26% | 3.63s |
| fetch (duckdb) | 2.72s | 14.98% | 6.35s |
| feature.setFields | 0.98s | 5.40% | 7.33s |
| QgsGeomettry creation + feature.setGeometry | 3.44s | 18.94% | 10.77s |
| feature.setId | 0.36s | 1.98% | 11.13s  |
| feature.setAttributes | 7.03s | 38.71% | 18.16s |

In other words, this means that this simple python code takes 7.33 seconds:

```py
feature = QgsFeature()
for _ in range(1_607_055):
    feature.setAttributes([42, "foo", "bah", "hello world", "QGIS"])
```
Could we do better? I hope so. 

### The setAttributes performance improvement

On the c++ side, a `QgsAttributes` is nothing more than a List of `QVariant` and the sip call to instantiate it just iterates the python list and calls `sipConvertToType`. This call takes a lot of time. Based of some investigation it looks this function needs to make an intermediate allocation which slows things down a lot.   
It turns out, that in some cases, on can simply avoid the `sipConvertToType`. The idea is to check the type of the input `attribute`. If this is a boolean, an integer, a float or a string, one can directly call the appropriate function of the python c API.

With this change, the previous `setAttributes` call now takes only 2s. A **72.7%** increased performance. Not bad. Not bad at all :-)

### Let's improve setAttribute too

The python `QgsFeature.setAttribute` has the same issue. So, I have made a similar change and I have measured up to **80%** gain.  Yeah :-)     
Moreover, there is no loss of performance if the attribute is already a `QVariant`. I have just measured how much times it takes to call 1_607_055 times `setAttribute`

|  | before | after|
| ------ | ------ |------ |
| integer |  1.71 s | 0.34 s |
| boolean |  1.70 s | 0.29 s |
| float |  1.79 s | 0.42 s |
| string |  2.07 s | 0.59 s |
| QVariant |  2.68 s | 2.68 s |

### The Qt6 situation 

The tests were done with the Qt5 version of QGIS in release mode. I have tried both sip4 and sip6. I got similar results. It looks like the sip version has no effect on the performance. 

I have also tested the Qt6 version of QGIS. It looks like Qt6 gained some performance to handle `QVariant`. However, one call still measure a lot nice improvements: 

|  | before | after|
| ------ | ------ |------ |
| integer |  0.81 s | 0,36 s |
| boolean |  0.76 s | 0.31 s |
| float |  0.90 s | 0.44 s |
| string |  1.28 s | 0.82 s |
| QVariant |  1.39 s | 1.39 s |
| setAttributes (1 integer et 4 strings) | 3.12 s | 1.04 s |

I am well aware that this PR changes some critical part of QGIS core source code. However, I think that the new performance is quite promising. I have tried all the unit test and I haven't noticed any regression. What do you think? Suggestions welcome!

### CODE REVIEW

File: python/PyQt6/core/auto_generated/qgsattributes.sip.in [hunk 1]
```diff
@@ -81,6 +81,22 @@ typedef QVector<QVariant> QgsAttributes;
     {
       qv->append( QVariant( QVariant::Int ) );
     }
+    else if ( PyBool_Check( obj ) )
+    {
+      qv->append( QVariant( PyObject_IsTrue( obj ) ) );
```
Comment (nyalldawson) [inline python/PyQt6/core/auto_generated/qgsattributes.sip.in:86]: https://docs.python.org/3/c-api/object.html#c.PyObject_IsTrue returns an integer value, but we need a bool here:

```suggestion
      qv->append( QVariant( PyObject_IsTrue( obj ) == 1 ) );
```

(Guess there's some test gaps here!)

File: python/PyQt6/core/auto_generated/qgsfeature.sip.in [hunk 2]
```diff
@@ -307,13 +307,43 @@ Alternatively, in Python it is possible to directly set a field's value via the
 %MethodCode
     bool rv;
 
-    if ( a1Wrapper == Py_None )
+    if ( a1 == Py_None )
     {
       rv = sipCpp->setAttribute( a0, QVariant( QVariant::Int ) );
     }
+    else if ( PyBool_Check( a1 ) )
```
Comment (nyalldawson) [inline python/PyQt6/core/auto_generated/qgsfeature.sip.in:314]: This optimisation could also be made for the setitem methods, and for the other setAttribute overloads

**Review by nyalldawson — COMMENTED** [review]
Fantastic, great optimisation ! 🥳 I can't see any issues with this approach.

**Comment by nyalldawson** [comment]
FYI the kdab hotspot tool is super useful in identifying areas for optimisation like this

**Comment by nyalldawson** [comment]
@ptitjano There's likely some additional speed ups you'll gain from https://github.com/qgis/QGIS/pull/57482

File: python/PyQt6/core/auto_generated/qgsattributes.sip.in (hunk 1 above)
Comment (ptitjano) [reply python/PyQt6/core/auto_generated/qgsattributes.sip.in]: Good catch. Fixed.

File: python/PyQt6/core/auto_generated/qgsfeature.sip.in (hunk 2 above)
Comment (ptitjano) [reply python/PyQt6/core/auto_generated/qgsfeature.sip.in:370]: This is now done.

**Comment by ptitjano** [comment]
I have added some tests to `test_qgsfeature.py` to improve the coverage.

**Comment by ptitjano** [comment]
> FYI the kdab hotspot tool is super useful in identifying areas for optimisation like this

That's good to know. Thanks.

File: tests/src/python/test_qgsfeature.py [hunk 3]
```diff
@@ -192,6 +212,73 @@ def test_setAttribute(self):
...
+            -9585674563452,
+            34.3,
+            False,
+            "QGIS",
+            QVariant("foo"),
+            QDate(2023, 1, 1),
+            QTime(12, 11, 10),
+            QDateTime(QDate(2020, 5, 6), QTime(8, 9, 10))
+        ]
+        self.assertTrue(feat.attributeCount() == 1)
```
Comment (nyalldawson) [inline tests/src/python/test_qgsfeature.py:226]:
```suggestion
        self.assertEqual(feat.attributeCount(), 1)
```

File: tests/src/python/test_qgsfeature.py [hunk 4]
```diff
@@ -192,6 +212,73 @@ def test_setAttribute(self):
...
+        for attribute in attributes:
+            self.assertTrue(feat.setAttribute(0, attribute))
+            self.assertTrue(feat.attribute(0) == attribute)
```
Comment (nyalldawson) [inline tests/src/python/test_qgsfeature.py:229]:
```suggestion
            self.assertEqual(feat.attribute(0), attribute)
```

File: tests/src/python/test_qgsfeature.py [hunk 5]
```diff
@@ -192,6 +212,73 @@ def test_setAttribute(self):
...
+
+            feat.setAttribute(0, None)
+            self.assertEqual(feat.attribute(0), None)
```
Comment (nyalldawson) [inline tests/src/python/test_qgsfeature.py:232]:
```suggestion
            self.assertIsNone(feat.attribute(0))
```

File: tests/src/python/test_qgsfeature.py [hunk 6]
```diff
@@ -192,6 +212,73 @@ def test_setAttribute(self):
...
+            fields.append(QgsField(attribute['name']))
+
+        feat = QgsFeature(fields)
+        feat.initAttributes(len(attributes))
+
+        for attr in attributes:
+            name = attr['name']
+            value = attr['value']
+            feat.setAttribute(name, value)
+            self.assertTrue(feat.attribute(name) == value)
```
Comment (nyalldawson) [inline tests/src/python/test_qgsfeature.py:274]:
```suggestion
            self.assertEqual(feat.attribute(name), value)
```

**Review by nyalldawson — APPROVED** [review]

File: tests/src/python/test_qgsfeature.py (hunk 4 above)
Comment (ptitjano) [reply tests/src/python/test_qgsfeature.py]: Fixed.

File: tests/src/python/test_qgsfeature.py (hunk 5 above)
Comment (ptitjano) [reply tests/src/python/test_qgsfeature.py]: Fixed.

File: tests/src/python/test_qgsfeature.py (hunk 6 above)
Comment (ptitjano) [reply tests/src/python/test_qgsfeature.py]: Fixed.

File: tests/src/python/test_qgsfeature.py (hunk 3 above)
Comment (ptitjano) [reply tests/src/python/test_qgsfeature.py]: Fixed.

File: tests/src/python/test_qgsfeature.py [hunk 7]
```diff
@@ -192,6 +212,80 @@ def test_setAttribute(self):
...
+        self.assertEqual(feat.attributeCount(), 1)
+        for attribute in attributes:
+            self.assertTrue(feat.setAttribute(0, attribute))
+            self.assertEqual(feat.attribute(0), attribute)
+
+            feat.setAttribute(0, None)
+            if int(QT_VERSION_STR.split('.')[0]) >= 6:
+                self.assertIsNone(feat.attribute(0))
+            else:
+                self.assertTrue(feat.attribute(0).isNull())
```
Comment (ptitjano) [inline tests/src/python/test_qgsfeature.py:235]: Qt 5 and Qt 6 do not handle "None" `QVariant` the same way. This has nothing to do with the change introduced by this MR. Do you think this is the right approach to test it?

Comment (nyalldawson) [reply tests/src/python/test_qgsfeature.py:235]: @ptitjano 

    self.assertEqual(feat.attribute(0), NULL)

should work in both versions

Comment (ptitjano) [reply tests/src/python/test_qgsfeature.py]: Done.

**Pull request merged**